crates/compiler/src/compiler.rs

```diff
@@ -156,6 +148,80 @@ fn add_tracking_declarations(mut state: CompilationIntermediate) -> CompilationI
     state
 }
 
+fn generate_code(mut state: CompilationIntermediate) -> CompilationIntermediate {
+    let has_errors = state
+        .diagnostics
+        .iter()
+        .any(|d| d.severity == DiagnosticSeverity::Error);
+    let results: Vec<_> = if has_errors {
+        // We have errors, so we can't safely generate code.
+        vec![]
+    } else {
+        // No errors! Go ahead and generate the code for all parsed
+        // files.
+        let template = CompilationResult {
+            string_table: state.string_table.0.clone(),
+            contains_implicit_string_tags: state.string_table.contains_implicit_string_tags(),
+            ..Default::default()
+        };
+        state
+            .parsed_files
+            .iter()
+            .map(|file| generate_code_for_file(&mut state.tracking_nodes, template.clone(), file))
+            .collect()
+    };
+    state.result = Some(CompilationResult::combine(
+        results,
+        state.string_table.clone(),
+    ));
+    state
+}
+
+fn generate_code_for_file<'a, 'b: 'a, 'input: 'a + 'b>(
+    tracking_nodes: &mut HashSet<String>,
+    result_template: CompilationResult,
+    file: &'a FileParseResult<'input>,
+) -> CompilationResult {
+    let compiler_listener = Box::new(CompilerListener::new(tracking_nodes.clone(), file.clone()));
+    let compiler_tracking_nodes = compiler_listener.tracking_nodes.clone();
+    let compiler_diagnostics = compiler_listener.diagnostics.clone();
+    let compiler_program = compiler_listener.program.clone();
+    let compiler_debug_infos = compiler_listener.debug_infos.clone();
+
+    YarnSpinnerParserTreeWalker::walk(compiler_listener, &*file.tree);
+
+    tracking_nodes.extend(compiler_tracking_nodes.borrow().iter().cloned());
+
+    // Don't attempt to generate debug information if compilation
```

Comment:
Make this one line as well

Reply:
[`135705d` (#56)](https://github.com/Mafii/rusty-yarn-spinner/pull/56/commits/135705d08c9f00f30e7f7cc3575646807a2ed703)